### PR TITLE
KAFKA-12773: Use UncheckedIOException when wrapping IOException

### PR DESCRIPTION
**Pull request by socutes**
The raft module may not be fully consistent on this but in general in that module we have decided to not throw the checked IOException. We have been avoiding checked IOException exceptions by wrapping them in RuntimeException. The raft module should instead wrap IOException in UncheckedIOException. This change should be limited to the raft module.

### CODE REVIEW

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (showuon) [inline raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java]: The indent is not consistent with others. Also, should we break the exception into 6 lines? I think this could be enough:
```java
    throw new UncheckedIOException(
            String.format("Read the Quorum status exception from the file %s", file), e);
```

Same comments to other places.

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (showuon) [inline raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java:71]: Should we replace `throws IOException` with `throws UncheckedIOException` here?

Same comments to other places.

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (showuon) [reply raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java]: Also, could you re-phrase the error message, ex: `Error while reading the Quorum status from the file`

Same comments to the other places.

**Review by showuon — COMMENTED** [review]
@socutes , thanks for the PR. Left some comments. Thanks.

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java]: Thank you for your review. It has been modified, please Review it again !

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java:71]: The exception has already been caught and handled. Do you need to throw it up?

Comment (showuon) [reply raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java:71]: You're right! Not necessary. Please ignore my comment. Thanks.

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (showuon) [inline raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java]: there should be a space after a comma:
```
String.format("Error while reading the Quorum status from the file %s", file), e);
```

Also, the indent is not correct here. Please fix it. And same to other places.
Thanks.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java
Comment (showuon) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: indent is not correct here.

File: raft/src/main/java/org/apache/kafka/raft/QuorumState.java
Comment (showuon) [inline raft/src/main/java/org/apache/kafka/raft/QuorumState.java]: Why should we change `IOException` into `Exception` here?

File: raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java
Comment (showuon) [inline raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java:950]: Why should we remove these 2 lines? After this PR, we should throw `UncheckedIOException` in corrupted store, is that right?

**Review by showuon — COMMENTED** [review]
@socutes , thanks for the update. Left some comments. Thanks.

File: raft/src/main/java/org/apache/kafka/raft/QuorumState.java
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/raft/QuorumState.java:120]: Previously, the IOException would have been thrown here, so you need to catch it here. After the KIP handles the IOException, the underlying IOException will not be thrown up, but will become other exceptions, such as UnCheckeDioException, so you need to handle it this way.

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java]: Sorry, the code format problem, will pay attention later, this time rushed submission

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: Newly added code, indentation seems to be correct?

File: raft/src/main/java/org/apache/kafka/raft/QuorumState.java
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/raft/QuorumState.java]: You're right, there needs to be a change to handle  UncheckedIOException.

Comment (showuon) [reply raft/src/main/java/org/apache/kafka/raft/QuorumState.java]: Yes, but I think `Exception` is covering too many "unexpected" exceptions. You can see the catch block, we are handling IOException case, not other exceptions. I think we can directly catch `UncheckedIOException` here.

File: raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java
Comment (socutes) [reply raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java:950]: You were right. I fixed it

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java
Comment (showuon) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: No, you can see the change, you move this line `if (quorum.isVoter()` left (4 spaces I guess), so the following 2 lines should also move left.

Comment (socutes) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: Thank you very much for your precious time. I understand it and I will fix it

File: raft/src/main/java/org/apache/kafka/raft/QuorumState.java
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/raft/QuorumState.java]: Yes, I seem to have

**Comment by socutes** [comment]
@showuon Please review the changes again! Thanks.

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (showuon) [inline raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java]: Maybe this is better: `Error while deleting file...`

File: raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java
Comment (showuon) [inline raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java:951]: `nit`: I think it'd better to put the 2 line mock at the beginning of the test as before. (i.e. before `QuorumState state = buildQuorumState(Utils.mkSet(localId));`)

**Review by showuon — COMMENTED** [review]
Left 2 comments. Thanks.

**Comment by socutes** [comment]
> Left 2 comments. Thanks.

fixed! @showuon Thanks.

**Comment by socutes** [comment]
> > Left 2 comments. Thanks.
> 
> fixed! @showuon Thanks.

File: raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java
Comment (socutes) [reply raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java:951]: @showuon Please review the changes again! Thanks.

**Review by showuon — APPROVED** [review]
LGTM! @socutes , thanks for the PR!

**Comment by showuon** [comment]
@jsancio @mumrah , could you have another review for this PR since you are more familiar with this ticket. Thank you.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:384]: This is semantically different from the previous code but I think it is okay. If `initialize` throws an io exception the JVM process for the broker will terminate as this is called in the `RaftManager` constructor which is indirectly called from `kafka.Kafka`'s `main`.

File: raft/src/main/java/org/apache/kafka/raft/QuorumState.java
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/QuorumState.java]: How about "Failed to transition from ${this.state.name()} to ${state.name()}"? Note that this string is  not valid Java.

**Review by jsancio — COMMENTED** [review]
Thanks for the changes. Looks good in general. I found a few other places:
1. `Snapshots.createTempFile`
2. `FileRawSnapshotWriter`
3. `FileRawSnapshotReader`
4. `RecordsIterator::readFileRecords`

**Comment by socutes** [comment]
@jsancio Please review the changes again! Thanks.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:384]: This change doesn't look like it needs to be changed, does it?

Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:384]: Yeah, I think it is fine as is. I was just pointing out that it is technically different from the original code.

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotReader.java
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotReader.java]: These 5 lines have extra 4 spaces of indentation.

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotReader.java
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotReader.java]: These 2 lines have extra 4 spaces of indentation.

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java]: Extra spaces of indentation. Even for method parameters, we only indent by 4 spaces. Also, you can clean this up if each argument for `String.format` is on its own line. E.g.
```java
String.format(
    "....",
    tempSnapshotPath,
    snapshotId
)
```

Note, this comment applies to a few other places in this file.

File: raft/src/main/java/org/apache/kafka/snapshot/Snapshots.java
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/snapshot/Snapshots.java]: Some of these lines have extra 4 space. Other lines have extra 8 spaces.

**Review by jsancio — APPROVED** [review]
LGTM. Just a few formatting suggestions.

**Comment by socutes** [comment]
@jsancio Code format fix completed! Thank you very much for your Review.

**Comment by socutes** [comment]
@jsancio @showuon Can this PR be merged into the trunk? Is there anything else I need to do?

**Comment by jsancio** [comment]
@mumrah or @hachikuji Can you take a look at this PR?

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java]: nit: indentation

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java]: nit: indentation

File: raft/src/main/java/org/apache/kafka/raft/QuorumState.java
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/raft/QuorumState.java]: nit: indentation

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotReader.java
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotReader.java]: nit: seems like this could all fit on one line without going over 120 characters

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java]: nit: same formatting comment as above

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java]: nit: indentation

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java]: nit: indentation

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java]: nit: indentation

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java]: nit: indentation

File: raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java]: nit: indentation

**Review by mumrah — CHANGES_REQUESTED** [review]
Lots of little formatting nits, otherwise this is looking good. Possibly some things you've already fixed got unfixed after merging in trunk.

**Comment by socutes** [comment]
@mumrah Code format fix completed! Thank you very much for your Review.

Before I submit the code, I execute the command "./gradlew checkstyleMain checkstyleTest "to verify the code style. To ensure that the code format conforms to the standard.

As for the code format, do you have any documents and tools to suggest?

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java
Comment (hachikuji) [inline raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java]: nit: missing a space after the comma. Also, it would be nice to be consistent with the use of comma vs period for the similar cases in this file

File: raft/src/main/java/org/apache/kafka/snapshot/Snapshots.java
Comment (hachikuji) [inline raft/src/main/java/org/apache/kafka/snapshot/Snapshots.java]: nit: couldn't we return here and remove `tempFile`?

**Review by hachikuji — COMMENTED** [review]
Thanks for the patch. Had just a couple nitpicks.

File: raft/src/main/java/org/apache/kafka/snapshot/Snapshots.java
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/snapshot/Snapshots.java]: You're right！Thanks

File: raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java
Comment (socutes) [reply raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java]: Fixed!

**Comment by socutes** [comment]
@hachikuji Please review the changes again! Thanks.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java
Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:384]: Leaving this note for future readers. My comments above are not accurate. I misread the diff generated by GitHub. When I wrote the comment, I was under the impression that the old code was handling, wrapping and re-throwing the `IOException`.
Instead the old code wrapped and re-threw the `IOException`; it was not handling the exception.

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the patch!

**Pull request merged**
